crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheChecker.java

```diff
@@ -0,0 +1,131 @@
+import java.time.Instant;
+import java.util.Optional;
+import java.util.Set;
+
+/** Checks if cached data is outdated. */
+public class CacheChecker {
+
+  private final Cache cache;
+
+  /**
+   * @return the last modified time for the file at {@code path}. Recursively finds the most recent
+   *     last modified time for all subfiles if the file is a directory.
+   */
+  private static FileTime getLastModifiedTime(Path path) throws IOException {
+    FileTime lastModifiedTime = Files.getLastModifiedTime(path);
+
+    if (Files.isReadable(path)) {
+      try {
+        Optional<FileTime> maxLastModifiedTime =
+            Files.walk(path)
+                .map(
+                    subFilePath -> {
+                      try {
+                        return Files.getLastModifiedTime(subFilePath);
+
+                      } catch (IOException ex) {
+                        throw new UncheckedIOException(ex);
+                      }
+                    })
+                .max(FileTime::compareTo);
+
+        if (!maxLastModifiedTime.isPresent()) {
+          throw new IllegalStateException(
+              "Could not get last modified time for all files in directory '" + path + "'");
+        }
+        if (maxLastModifiedTime.get().compareTo(lastModifiedTime) > 0) {
+          lastModifiedTime = maxLastModifiedTime.get();
+        }
+
+      } catch (UncheckedIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    return lastModifiedTime;
+  }
+
+  public CacheChecker(Cache cache) {
+    this.cache = cache;
+  }
+
+  /** @return true if the base image layers exist in the cache; false otherwise */
+  public boolean areBaseImageLayersCached(ImageLayers<ReferenceLayer> baseImageLayers)
+      throws CacheMetadataCorruptedException {
+    // Gets the base image layers.
+    ImageLayers<CachedLayerWithMetadata> cachedLayers =
+        cache.getMetadata().filterLayers().byType(CachedLayerType.BASE).filter();
+
+    if (cachedLayers.size() < baseImageLayers.size()) {
+      return false;
+    }
+
+    for (ReferenceLayer baseImageLayer : baseImageLayers) {
+      if (!cachedLayers.has(baseImageLayer.getBlobDescriptor().getDigest())) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Checks all cached layers built from the source files to see if the source files have been
+   * modified since the newest layer build.
+   *
+   * @param sourceFiles the source files to check
+   * @return true if no cached layer exists that are up-to-date with the source files; false
+   *     otherwise.
+   */
+  public boolean areSourceFilesModified(Set<Path> sourceFiles)
+      throws IOException, CacheMetadataCorruptedException {
+    // Grabs all the layers that have matching source files.
+    ImageLayers<CachedLayerWithMetadata> cachedLayersWithSourceFiles =
+        cache.getMetadata().filterLayers().bySourceFiles(sourceFiles).filter();
+    if (cachedLayersWithSourceFiles.size() == 0) return true;
```

Comment:
Google style is to never use single line if:

https://google.github.io/styleguide/javaguide.html#s4.1-braces